autopilot/agent.go

```diff
@@ -578,18 +579,58 @@ func (a *Agent) openChans(availableFunds btcutil.Amount, numChans uint32,
 	)
 	a.pendingMtx.Unlock()
 
-	// If we reach this point, then according to our heuristic we
-	// should modify our channel state to tend towards what it
-	// determines to the optimal state. So we'll call Select to get
-	// a fresh batch of attachment directives, passing in the
-	// amount of funds available for us to use.
-	chanCandidates, err := a.cfg.Heuristic.Select(
-		a.cfg.Self, a.cfg.Graph, availableFunds,
-		numChans, nodesToSkip,
+	// Gather the set of all nodes in the graph, except those we
+	// want to skip.
+	selfPubBytes := a.cfg.Self.SerializeCompressed()
+	nodes := make(map[NodeID]struct{})
+	if err := a.cfg.Graph.ForEachNode(func(node Node) error {
+		nID := NodeID(node.PubKey())
+
+		// If we come across ourselves, them we'll continue in
+		// order to avoid attempting to make a channel with
+		// ourselves.
+		if bytes.Equal(nID[:], selfPubBytes) {
+			return nil
+		}
+
+		// Additionally, if this node is in the blacklist, then
+		// we'll skip it.
+		if _, ok := nodesToSkip[nID]; ok {
+			return nil
+		}
+
+		nodes[nID] = struct{}{}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("unable to get graph nodes: %v", err)
+	}
+
+	// Use the heuristic to calculate a score for each node in the
```

Comment:
I dig the control flow of this new loop, haven't got to the rest of the commits yet, but as joost suggested in a prior commit, we can make a composite heuristic that itself combines other heuristics and allows us to hide the details w.r.t the eventual normalization scoring we add in.